okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -721,8 +732,14 @@ void finishStream(
           frameWriter.rstStream(streamId, ErrorCode.CANCEL);
         }
         if (status != null) {
-          stream.transportState().transportReportStatus(status, stopDelivery,
-              trailers != null ? trailers : new Metadata());
+          boolean isCancelled =
+              (status.getCode() == Code.CANCELLED || status.getCode() == Code.DEADLINE_EXCEEDED);
+          stream
+              .transportState()
+              .transportReportStatus(
+                  status,
+                  isCancelled || stopDelivery,
```

Comment:
The stop on CANCELLED or DEADLINE_EXCEEDED is the existing logic here. I should not have removed this entirely in part 3 of this PR - I now am just adding it back to preserve the current behavior, alongside the now-necessary explicit stopDelivery flag. 

Do you want this logic moved to the rst stream method? I was trying to minimize changes to other code paths, which would also potentially be impacted by moving the isCancelled check - this might be the appropriate place for the check, but didn't seem relevant to the changes here.